File: Leetcode/Medium/1400.java
// https://leetcode.com/problems/construct-k-palindrome-strings/description/?envType=daily-question&envId=2025-01-11

class Solution {
    public boolean canConstruct(String s, int k) {
        if (s.length() < k) return false; // string not long enough
        if (s.length() == k) return true; // k 1-char strings
        int[] freq = new int[26];
        for (char c : s.toCharArray()) freq[c - 'a']++;
        // if X is the number of odd chars, we must make at least X palindromic strings
        int odd = 0;
        for (int x : freq) odd += (x & 1);
        return (odd <= k); // ??
        /*
        because of the checks with s.length(), we guaranteed that
        we make 'odd' strings with the odd-frequency characters
        and the other 'k - odd' strings with the even-frequency chars
        */
    }

    public boolean alt(String s, int k) {
        // this uses an integer to store the frequencies,
        // and the number of 1s is the number of odd-freq
        if (s.length() < k) return false; // string not long enough
        if (s.length() == k) return true; // k 1-char strings
        int tmp = 0;
        for (char c : s.toCharArray()) tmp ^= 1 << (c - 'a');
        return Integer.bitCount(tmp) <= k;
    }
}

File: Leetcode/Medium/1400.c
// https://leetcode.com/problems/construct-k-palindrome-strings/?envType=daily-question&envId=2025-01-11

bool canConstruct(char* s, int k) {
    // WOW c is a very weird language
    // ignore the fact this is like terrible language usage
    int* n = (int*) malloc(sizeof(int)); *n = strlen(s);
    if (*n <= k) {
        bool ans = (*n == k);
        free(n);
        return ans;
    }
    bool* freq = (bool*) calloc(26, sizeof(bool));
    for (int i = 0; i < *n; i++) {
        freq[s[i] - 'a'] ^= 1;
    }
    free(n);
    int* odd = (int*) malloc(sizeof(int)); *odd = 0;
    for (int i = 0; i < 26; i++) *odd += freq[i];
    bool ans = (*odd <= k);
    free(freq);
    free(odd);
    return ans;
}

bool alt(char* s, int k) {
    // this is more real-person code i think (bcz technically you shouldn't malloc everything LOL)
    int n = strlen(s);
    if (n <= k) return (n == k);
    bool* freq = (bool*) calloc(26, sizeof(bool));
    for (int i = 0; i < n; i++) {
        freq[s[i] - 'a'] ^= 1;
    }
    int odd = 0;
    for (int i = 0; i < 26; i++) odd += freq[i];
    free(freq);
    return (odd <= k);
}